### DISCUSSION TITLE
Welcome to kyberJCE Discussions!

### DISCUSSION THREAD

**Discussion by fisherstevenk** [fisherstevenk/kyberJCE#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by EileenVonEdel** [comment]
Good Day to you,

i was wondering if you have an idea how to fix an error that occurs when i'm using your implementation. I'm a student and for my bachelor thesis i'm working on an post-quantum secure messaging android app. For this i wanted to support two different Post-Quantum Encryption Methods. NTRU and Crystals Kyber. Within my research i found your implementation which seems to be perfect for me to use. 
However when i add the jar to my android studio as a library and try to use it i always get the following exception.

`Caused by: java.lang.NoSuchFieldError: No field PROVIDER_VER of type Ljava/lang/String; in class Lsun/security/util/SecurityConstants; or its superclasses (declaration of 'sun.security.util.SecurityConstants' appears in /system/framework/core-oj.jar)
        at com.swiftcryptollc.crypto.provider.KyberJCE.<init>(KyberJCE.java:47)`

I don't understand why this is happening. The error occurs when im calling new KyberJCE().
I am using the newest Android Studio Version which uses JDK 11 as default. It kind of confuses me because i thought that the class SecurityConstants is part of the JDK but i can't import any sun package.
I already tried to use different JDKs ,especially OpenJDK, but nothing worked it's always the same error. 

I made sure to specify the compileVersion to JAVA_11 but could it be that i missed to configure something else in Android Studio?

I would be very happy if you have some ideas how to fix this error.

Best Wishes
Eileen von Edel

**Reply by fisherstevenk** [reply]
Hello. How are you?  Someone else just asked me the same thing.  I'll have to add this into the readme to make it easier.

You can find the source code for OpenJDK 11 and 13 here with those classes.  The JDK build for your Android doesn't have those classes included.

https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/sun/security/util

https://github.com/openjdk/jdk/tree/jdk-13%2B33/src/java.base/share/classes/sun/security/util

You could try to take the sun classes and compile them into a jar. (I haven't done this, so I don't know if it'll work off the top of my head without including other libraries. )

https://github.com/openjdk/jdk/tree/jdk-11+28/src/java.base/share/classes/sun

https://github.com/openjdk/jdk/tree/jdk-13%2B33/src/java.base/share/classes/sun

There is also a "fat" jar which contains all the libraries you need for your code.  You could compile the KyberJCE jar outside of Android with a full OpenJDK and then import that jar into Android Studio.  (I'm not sure how big that jar would be though.)

https://stackoverflow.com/questions/19150811/what-is-a-fat-jar#29925421

https://dzone.com/articles/the-skinny-on-fat-thin-hollow-and-uber

Hopefully one of those things helps you; if not, let me know.  Good luck!

V/r,
Steve

**Comment by YYZStart** [comment]
Greetings ,

I am wondering if you have any idea of the error when I implement your work . I'm an undergraduate student and I'm working on my bachelor thesis ,which is also for secure messaging .

I'm using Eclipse Version: 2022-03 (4.23.0) with JDK18 to develop my work. I download and unzip your Source code implementation of KyberJCE from Github, then I import your unzip Source code as the existing Maven project into Eclipse. 
An error occurred: "import sun. security. util": `The type sun. Security. Util. DerInputStream is not accessible.`

As your reply to  EileenVonEdel , another disscussion above , you pointed out that Android doesn't have those classes included.
I'm using Eclipse , I follow your instruction to build a Sun.jar which gets from openJDK13 ,then I made it build path .
However , the import sun.security.util error still existed .

Also , I tried to modified the Access Rule by "Properites->Java Build Path-Libraries, so as to make sun.security.util accessible , but error still existed .

So I tried another solution : Build a new maven project and make your kyberJCE-2.1.1.jar build path ,following your pom file rule , howerver , another error occurred when calling "Security.addProvider(new KyberJCE());".

```
Exception in thread "main" java.lang.IllegalAccessError: class com.swiftcryptollc.crypto.provider.KyberJCE (in unnamed module @0x3a4afd8d) cannot access class sun.security.util.SecurityConstants (in module java.base) because module java.base does not export sun.security.util to unnamed module @0x3a4afd8d
at com.swiftcryptollc.crypto.provider.KyberJCE.<init>(KyberJCE.java:48)
```


I'm afraid that there're errors in my pom file , so I use your Source code file to do above things again ,but ended up the same errors...
I've tried changing the Eclipse version even install another JDK version , the error still existed.

I would appreciate if you have some ideas how to fix this error.

Sorry for my bad English, the grammar or the words may look weird...


Sincerely ,

LenghuoTrain

**Reply by fisherstevenk** [reply]
Hello.   For JDK 18, you need to add 2 things to your pom file.  (Please note that this project does have it's own pom file with these 2 things in it..)

<project ...>
**<properties>
     
     <argLine>--add-modules java.base --add-opens java.base/sun.security.util=ALL-UNNAMED</argLine>
   </properties>**

And then you need to modify the maven compiler plug-in:

<build>
     <plugins>
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
         <version>3.10.1</version>
         <configuration>
           <source>18</source>
           <target>18</target>
           **
           <compilerArgs>
             <arg>--add-exports</arg>
             <arg>java.base/sun.security.util=ALL-UNNAMED</arg>
           </compilerArgs>**
         </configuration>
       </plugin>


Note that you can compile it with the OracleJDK, but running it in another app, you'll need the OpenJDK (or similar) since your Jar won't be signed. 

You could use the 2.1.1 jar in the builds directory if you just want to push forward.  That one is signed (I applied to Oracle for a JCE signing certificate) and you'll be able to import it into the OracleJDK.

I hope that's the missing piece for you.  If you need more help with it, please let me know and I'll see if I can help. 

V/r,
Steve

**Comment by ronhombre** [comment]
Hello!

I was gonna implement Kyber in pure Java. Good thing I checked first. Have you cross-checked the outputs of this implementation with others? Like the ones made by Amazon and Cloudflare.

I'm wondering because I'm planning to implement Kyber into an app I'm developing.

Hopeful,
Ron

**Reply by fisherstevenk** [reply]
Hello.

It was not compared against Amazon or Cloudflare.  This project was based on the Go implementation (from a third party) and C implementation from the authors of Kyber.  Their test code was converted to Java and the test inputs were also brought in.

The authors reviewed this implementation and you can find this repo is listed on the CRYSTALS Kyber software page by the authors.

https://pq-crystals.org/kyber/software.shtml

-Steve

**Reply by ronhombre** [reply]
Thank you for the reply! This was a rather old question. I have made my own implementation on Kotlin based on FIPS 203 ipd. I understand there is still ambiguity at the moment whether the pseudocode in FIPS 203 should be followed because it produces values different from Kyber v3. There are two places where the matrix A was transposed.

**Comment by mohamed-amine-mighri** [comment]
Hello,

I hope this message finds you well.

I was hoping to seek your assistance regarding a matter. I am currently a student working on my bachelor's thesis, which revolves around a post-quantum IoT project. Specifically, I am endeavoring to implement KyberJCE for securing messages within this project. However, I have encountered an issue. It appears that the public keys and private keys generated differ from the specifications outlined by NIST.

Upon comparison, I have observed that the private keys are consistently larger by 169 bytes than the size stipulated by NIST. Similarly, the public keys exhibit a discrepancy, being consistently larger by 167 bytes than the NIST specifications dictate.

Your insights and guidance on this matter would be greatly appreciated.

Thank you for your attention to this inquiry.

Best regards,

**Reply by ronhombre** [reply]
I believe Steven is busy. You can use my implementation, but it's not truly based on _FIPS 203 ipd_. The Kyber community has unanimously decided to stay aligned with the _Version 3.02_ paper instead of the NIST paper. The only difference is that two matrices are transposed the other way. Don't worry. The NIST paper is a draft while the Version 3.02 has been implemented by a lot of people before the NIST paper came out, and there is still no clear answer about the discrepancy to the best of my knowledge. Basically, we decided against the NIST implementation because it's a draft and for backward compatibility with previous Kyber implementations. I'm currently waiting for the next FIPS 203 draft to see if I keep the transposed matrices or not.

[KyberKotlin](https://github.com/ronhombre/KyberKotlin)
